ball_finder.py

```diff
@@ -0,0 +1,81 @@
+import cv2
+import math
+import numpy as np
+
+
+class BallFinder:
+    def __init__(self, detector):
+        self.detector = detector
+        self.generator = detector.get_generator()
+
+        frame, _ = self.generator.generate()
+        self.SCREEN_HEIGHT, self.SCREEN_WIDTH = frame.shape[:2]
+        self.FOCAL_LENGTH_PIXELS = 657
+
+        # Depth csv of empty floor
+        self.floor_frame = np.loadtxt("FLOOR.csv", dtype=np.float32, delimiter=",")
+        self.floor_frame[self.floor_frame > 4] = 0
+
+        # Displacement constants between rgb frame and depth frame
+        self.DEPTH_X_SHIFT = -23
+        self.DEPTH_Y_SHIFT = 31
+
+        # Height of the camera (in m)
+        self.CAMERA_HEIGHT = 0.7
+
+        # Minimum distance (in m) for finding obstacles (the distance from the camera to the front of the robot)
+        self.MIN_OBSTACLE_DIST = 0.56
+
+    # Returns depth frame where all floor points have depth 0.
+    def remove_floor(self, depth):
+        return np.where(abs(depth - self.floor_frame) < 0.05, 0, depth)
+
+    # Returns the center (x, y) of a contour based on its bounding box.
+    def get_contour_center(self, contour):
+        x, y, w, h = cv2.boundingRect(contour)
+        return x + w / 2, y + h / 2
+
+    # Returns horizontally straight distance (in m) to a ball or obstacle.
+    # is_ball is used to determine whether or not to shift the x and y positions the amount needed when using the depth
+    # camera.
+    def get_distance(self, depth_frame, item, is_ball=False):
+        x = min(639, int(item[0] + is_ball * self.DEPTH_X_SHIFT))
+        y = min(479, int(item[1] + is_ball * self.DEPTH_Y_SHIFT))
+        return math.sqrt(math.pow(depth_frame[y, x], 2) - math.pow(self.CAMERA_HEIGHT, 2))
+
+    # Returns angle (in degrees) between center of camera to center of ball.
+    def get_angle_deg(self, ball):
+        dist_to_center = ball[0] - self.SCREEN_WIDTH / 2
+        return math.degrees(math.atan(dist_to_center / self.FOCAL_LENGTH_PIXELS))
+
+    # Returns list of obstacle contours with area greater than 1000 pixels.
+    def get_obstacles(self, depth, max_distance):
+        if depth is None:
+            return []
+
+        mask = cv2.inRange(self.remove_floor(depth), self.MIN_OBSTACLE_DIST, float(max_distance) - 0.05)
+        obstacle_contours, _ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+        obstacle_contours.sort(key=lambda obstacle: self.get_distance(depth, self.get_contour_center(obstacle)))
```

Comment:
Basic question: Does it really make a huge difference to calculate the distance on the floor vs. just sorting by the depth to the camera? If we make the assumption that balls are on the floor, is there a situation where an obstacle appears closer to the camera than a ball but isn't blocking our path to the ball?

Reply:
I'm not completely sure what you mean by the first question. But for the second question, there could be a situation like that, but what we could do is provide them with more data than a boolean true or false(like a distance and angle) and have them decide what they'd like to do with that data.